Add file-descriptor I/O functions

madroach opened the pull request:
Beware that this has only been tested on OpenBSD.

Commit:
Add file-descriptor I/O functions

c-cube commented:
This is bigger than the other PRs so I'll look after they're merged. I think it also deserves to have a proper test suite, given that there's non trivial C stubs :-)

Commit:
Fix careless merge: Bigstring.t -> t

Commit:
add tests for Bigstring_unix.{read,write}

Commit:
remove verbose bounds-check inerited from Bigstringaf

Commit:
fix automatic length calculation when offset is provided

this bug was exposed by the added tests.

Commit:
add test-dependency on alcotest

Commit:
Tests also need opam depend on bigstring

My dune finds the bigstring library in the current dir.
Why the heck does the travis dune complain?

Commit:
bigstring_unix -> bigstring-unix

Commit:
Compatibility with OCaml <4.07 - exception patterns

Commit:
~cloexec only available since OCaml 4.05

Commit:
Tests also need opam depend on bigstring-unix

My dune finds the bigstring(-unix) library in the current workspace.
Why the heck does the travis dune complain?

c-cube commented on bigstring.opam:
```diff
@@ -13,9 +13,11 @@ build: [
   "base-bytes"
   "ocaml" {>= "4.03.0"}
+  "alcotest" {with-test}
+  "bigstring-unix" {with-test}
```
this seems unecessary? tests for bigstring-unix should be in the bigstring-unix package :-)

madroach replied:
Should I try to disable tests for ``bigstring`` ? But why care?

c-cube commented:
Tests are passing, I'll review the code soon. Thanks a lot :)

c-cube commented on src/bigstring_unix.ml:
```diff
@@ -3,6 +3,29 @@
+  then invalid_arg ("Bigstring_unix." ^ name)
+  else (off, len)
+
+external read_fd  : Unix.file_descr -> t -> int -> int -> int = "bigstring_read"
```
is there a risk of collision there? could it be something like `ocaml_bigstring_unix_read`, which is less likely to collide with other libraries?

madroach replied:
I have no idea. Let's rename it as you suggest.

c-cube commented on src/bigstring_unix.mli:
```diff
@@ -6,6 +6,15 @@ type t = (char, Bigarray.int8_unsigned_elt, Bigarray.c_layout) Bigarray.Array1.t
 
+(** {2 I/O} *)
+
+(** These I/O functions are missing from the Bigarray library.
```
not missing anymore, are they? :)

madroach replied:
They will still be missing from the ``Bigarray``, even when we add the to the ``Bigstring`` library. Should I remove this sentence?

c-cube commented on src/bigstring_unix_stubs.c:
```diff
@@ -0,0 +1,136 @@
+            err = len = 0;
+            break;
+          case ERROR_MORE_DATA:
+#if 0
```
why is that there? :)

madroach replied:
The ``ERROR_MORE_DATA``?
That's one thing where Windows behaves different then Unix for message based sockets. Unix discards the rest of the message when the provided buffer is too small, windows returns this error and will return the rest of the message in the next call. So the ``#if 0`` part would mimic Unix behaviour and discard the rest of the message. I'm not sure what to do on that error condition.

c-cube commented on src/bigstring_unix_stubs.c:
```diff
@@ -0,0 +1,136 @@
+CAMLprim value
+bigstring_read(value vfd, value vba, value voff, value vlen)
+{
+    void *iobuf = ((char *)Caml_ba_data_val(vba)) + Unsigned_long_val(voff);
```
There should be some `Caml_param` macro here, at least for `vfd` since it's a heap allocated object, right?

madroach replied:
I made sure not to dereference any pointer into the OCaml heap after ``caml_release_runtime_system()``. I don't do any allocations. So no need to register pointers into the OCaml heap with the Gc. That's the nice thing about bigarrays :)

c-cube commented on src/bigstring_unix_stubs.c:
```diff
@@ -0,0 +1,136 @@
+
+CAMLprim value
+bigstring_write(value vfd, value vba, value voff, value vlen)
+{
```
same here, seems like `vfd` should be protected with a caml macro.

madroach replied:
see above

c-cube commented on test/dune:
```diff
@@ -0,0 +1,11 @@
+ (flags :standard -thread)
+ (libraries alcotest bigstring bigstring-unix threads)
+ (action
+  (setenv MALLOC_CHECK_ 3
```
interesting, is that specific to a toolchain? a comment would be helpful .

madroach replied:
I will add a comment. This is enabled by default on OpenBSD and it helped me to discover some nasty bugs early. See http://man7.org/linux/man-pages/man3/mallopt.3.html#DESCRIPTION

c-cube commented on test/test.ml:
```diff
@@ -0,0 +1,83 @@
+    |> match len with None -> ignore |Some len -> check int "write length" len
+  in
+  "Bigstring_unix",
+  [ "read", `Quick, begin fun () ->
```
calls to `Gc.compact()` can also help to find potential memory issues.

madroach replied:
I added a call to ``Gc.compact ()`` to ``check_buf``.

c-cube reviewed (changes requested):
I'm a bit worried by the C bindings, they're really subtle to write, and I don't see the normal Caml macros.

Commit:
prefix c stubs with ocaml_bigstring_unix

Commit:
add Gc.compact to tests

Commit:
fix wrong variable references in Windows stubs

Commit:
explain MALLOC_PERTURB_

Commit:
add appveyor to test on Windows

c-cube commented:
tests seem to fail on linux. I trust you with merging that once tests pass, if you take ownership of that part of the code :-)

madroach commented:
That error is easy to fix. Since the test doesn't  ``exec`` I don't need to care about ``~cloexec``.

I cannot enable appveyor on this repository. Probably because I'm not the owner. Could you please enable appveyor? I do appveyor builds on my clone: https://ci.appveyor.com/project/madroach/ocaml-bigstring/builds/28734468

Commit:
remove ~cloexec, which is only available since OCaml 4.05

c-cube commented:
I think I enabled appveyor, can you confirm? :)

madroach commented:
Yes, confirmed. I guess it does not discover existing pull requests, but only starts builds on new ones. So just lets merge and see what happens :)

c-cube commented:
you can just push a change, too ;)

c-cube commented:
we should do a release once everything works.
